### PR TITLE
chapi: allow multiple targets per volume

### PR DESCRIPTION
**Pull request by shivamerla**
* Problem:
  * 3par has multiple target portals/iqn's for same volume
* Implementation:
  * allow multiple iqns per volume object and perform discovery/login accordingly
  * when virtual ip(discovery-ip) is provided single discovery-ip is used for discovery
* Testing:
* Review: gcostea, rkumar
* Bug: https://nimblejira.nimblestorage.com/browse/NLT-
Signed-off-by: Shiva Krishna, Merla <shivakrishna.merla@hpe.com>

### CODE REVIEW

File: linux/device.go [hunk 1]
```diff
@@ -449,8 +449,8 @@ func createLinuxDevice(volume *model.Volume) (dev *model.Device, err error) {
 	// cleanup paths which maybe part of lsscsi but not in multipath as we haven't found the device yet
 	cleanupStaleScsiPaths(volume)
 	// try to logout the iscsi target if we could not find a device by now for VST Volume
-	if volume.TargetScope != GroupScope.String() && volume.Iqn != "" {
-		iscsiLogoutOfTarget(&model.IscsiTarget{Name: volume.Iqn})
+	if volume.TargetScope != GroupScope.String() && len(volume.TargetNames()) != 0 {
+		iscsiLogoutOfTarget(&model.IscsiTarget{Name: volume.TargetNames()[0]})
```
Comment (rgcostea) [inline linux/device.go]: What if there are more than 1?  Don't we have to log out of all targets?

Comment (shivamerla) [reply linux/device.go]: no, we logout only for VST volumes. with GST as other devices are sharing it we never logout.

File: linux/iscsi.go [hunk 2]
```diff
@@ -95,6 +95,122 @@ func getByPathTargetPattern() string {
...
+func areTargetsLoggedIn(requiredTargets []string) (bool, error) {
+	loggedInTargets, err := GetLoggedInIscsiTargets()
+	if err != nil {
+		return false, fmt.Errorf("unable to determine already logged-in targets, err %s", err.Error())
+	}
+	for _, requiredTarget := range requiredTargets {
+		found := false
+		for _, loggedInTarget := range loggedInTargets {
+			if loggedInTarget == requiredTarget {
+				found = true
```
Comment (rgcostea) [inline linux/iscsi.go]: you can break here

File: linux/iscsi.go [hunk 3]
```diff
@@ -95,6 +95,122 @@ func getByPathTargetPattern() string {
...
+	if err != nil {
+		return err
+	}
+
+	// Get iscsi ifaces bound to network interfaces
+	ifaces, err := GetIfaces()
+	// treat iface path not found error as no ifaces bound
+	if err != nil && !os.IsNotExist(err) {
+		log.Errorf("Unable to retrieve iSCSI bound ifaces. Error: %s", err.Error())
+		return fmt.Errorf("Unable to retrieve iSCSI bound ifaces. Error: %s", err.Error())
```
Comment (rgcostea) [inline linux/iscsi.go]: lower case "U" to follow the pattern

File: linux/iscsi.go [hunk 4]
```diff
@@ -95,6 +95,122 @@ func getByPathTargetPattern() string {
...
+	// single-target-single-lun models doesn't require SCSI resan to be performed
+	if !strings.EqualFold(volume.TargetScope, GroupScope.String()) {
+		return nil
+	}
+
+	// perform SCSI rescan to create linux block devices
+	err = RescanIscsi(volume.LunID)
+	if err != nil {
+		log.Errorf("Unable to rescan iscsi hosts, Error: %s", err.Error())
+		return fmt.Errorf("Unable to rescan iscsi hosts, Error: %s", err.Error())
```
Comment (rgcostea) [inline linux/iscsi.go]: lowercase "U" here too

File: linux/iscsi.go (hunk 2 above)
Comment (shivamerla) [reply linux/iscsi.go]: sure..

File: linux/device.go (hunk 1 above)
Comment (rgcostea) [reply linux/device.go]: But won't this code be shared with 3Par?  Does that apply to them too?

Comment (shivamerla) [reply linux/device.go]: yes, 3par is GST by default. They should be setting TargetScope as group. They can be logged-in always without any devices and doesn't cause any issues in linux.

Comment (rgcostea) [reply linux/device.go]: Cool.  Thanks.

Comment (rgcostea) [reply linux/device.go]: BTW, they don't set target_scope.  It's assumed to be group if not provided.

Comment (shivamerla) [reply linux/device.go]: got it..yes that is good.

**Comment by shivamerla** [comment]
@wdurairaj can you review this as well? Also it will be great if you can test this branch with 3Par. I will cover all nimble tests before merge.

**Comment by rgcostea** [comment]
> lgtm. This would be a breaking change for other chapi consumers too (cv,flexvol etc). Are we planning to be backward compatible

Why is this breaking?  This code:

```
func (v Volume) TargetNames() []string {
	if v.Iqn != "" {
		return []string{v.Iqn}
	}
	return v.Iqns
}
```
Handles backwards compatibility.

**Comment by raunakkumar** [comment]
I misinterpreted target _name to be part of volume struct. This should work. Thanks

**Comment by shivamerla** [comment]
Tests with CSP changes:

```
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> stageVolume, volumeID: 0666c9b39533bd05d20000000000000000000009be, stagingMountPoint: /var/lib/kubelet/plugins/hpe.com/mounts/0666c9b39533bd05d20000000000000000000009be, volumeAccessType: mount, volCap: mount:<fs_type:\"xfs\" > access_mode:<mode:SINGLE_NODE_WRITER > , publishContext: map[accessProtocol:iscsi chapPassword: chapUsername: discoveryIps:172.16.34.159,172.16.226.159 fsCreateOptions: fsMode:0644 fsOwner:504:21 fsType:xfs lunId:1 readOnly:false serialNumber:ec40405760be73ff6c9ce90047b67455 targetNames:iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2 targetScope:group volumeAccessMode:mount], volumeContext: map[accessProtocol:iscsi fsMode:0644 fsOwner:504:21 fsType:xfs storage.kubernetes.io/csiProvisionerIdentity:1575928454982-8081-csi.hpe.com volumeAccessMode:mount]" file="node_server.go:364"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> setupDevice, publishContext: map[accessProtocol:iscsi chapPassword: chapUsername: discoveryIps:172.16.34.159,172.16.226.159 fsCreateOptions: fsMode:0644 fsOwner:504:21 fsType:xfs lunId:1 readOnly:false serialNumber:ec40405760be73ff6c9ce90047b67455 targetNames:iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2 targetScope:group volumeAccessMode:mount]" file="node_server.go:423"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> GetDeviceFromVolume for serial ec40405760be73ff6c9ce90047b67455" file="device.go:869"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>>> GetLinuxDmDevices called with ec40405760be73ff6c9ce90047b67455 and lunID 1" file="device.go:184"
time="2019-12-20T00:44:18Z" level=trace msg="ExecCommandOutputWithTimeout called with dmsetup[ls --target multipath] 60" file="cmd.go:35"
time="2019-12-20T00:44:18Z" level=trace msg="process with pid: 27 finished successfully" file="cmd.go:62"
time="2019-12-20T00:44:18Z" level=trace file="cmd.go:66"
time="2019-12-20T00:44:18Z" level=debug msg="Found 0 devices" file="device.go:321"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<<< GetLinuxDmDevices" file="device.go:327"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< GetDeviceFromVolume" file="device.go:877"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> CreateLinuxDevices" file="device.go:884"
time="2019-12-20T00:44:18Z" level=trace msg="create request with serialnumber :ec40405760be73ff6c9ce90047b67455, accessprotocol iscsi discoveryip , iqn  " file="device.go:889"
time="2019-12-20T00:44:18Z" level=debug msg=">>>> createLinuxDevice called with volume  serialNumber ec40405760be73ff6c9ce90047b67455 and lunID 1" file="device.go:407"
time="2019-12-20T00:44:18Z" level=trace msg="Called rescanVolume  and accessProtocol iscsi" file="device.go:378"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> HandleIscsiDiscovery for volume " file="iscsi.go:187"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> GetLoggedInIscsiTargets" file="iscsi.go:564"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> FileExists for path /sys/class/iscsi_session" file="file.go:62"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< FileExists" file="file.go:72"
time="2019-12-20T00:44:18Z" level=debug msg="found 0 logged-in targets" file="iscsi.go:596"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< GetLoggedInIscsiTargets" file="iscsi.go:597"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> loginToVolume for volume " file="iscsi.go:141"
time="2019-12-20T00:44:18Z" level=trace msg="PING  --> (172.16.34.159)" file="iscsi.go:362"
time="2019-12-20T00:44:18Z" level=trace msg="Received 24 bytes from 172.16.34.159: icmp_seq=0 time=1.437796ms ttl=64\n" file="iscsi.go:355"
time="2019-12-20T00:44:18Z" level=trace msg="PING  --> (172.16.226.159)" file="iscsi.go:362"
time="2019-12-20T00:44:18Z" level=trace msg="Received 24 bytes from 172.16.226.159: icmp_seq=0 time=549.762µs ttl=64\n" file="iscsi.go:355"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> PerformDiscovery with discovery IPs [172.16.34.159 172.16.226.159]" file="iscsi.go:721"
time="2019-12-20T00:44:18Z" level=trace msg="PING  --> (172.16.34.159)" file="iscsi.go:362"
time="2019-12-20T00:44:18Z" level=trace msg="Received 24 bytes from 172.16.34.159: icmp_seq=0 time=269.667µs ttl=64\n" file="iscsi.go:355"
time="2019-12-20T00:44:18Z" level=trace msg="ExecCommandOutputWithTimeout called with iscsiadm[-m discovery -t st -p 172.16.34.159 -o new] 60" file="cmd.go:35"
time="2019-12-20T00:44:18Z" level=trace msg="process with pid: 28 finished successfully" file="cmd.go:62"
time="2019-12-20T00:44:18Z" level=trace msg="172.16.226.158:3260,2461 iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2\n172.16.34.158:3260,2461 iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2\n" file="cmd.go:66"
time="2019-12-20T00:44:18Z" level=trace msg="172.16.226.158:3260,2461 iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2\n172.16.34.158:3260,2461 iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2\n" file="iscsi.go:754"
time="2019-12-20T00:44:18Z" level=trace msg="Name iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2 Address 172.16.226.158 Port 3260" file="iscsi.go:769"
time="2019-12-20T00:44:18Z" level=trace msg="Name iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2 Address 172.16.34.158 Port 3260" file="iscsi.go:769"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< PerformDiscovery" file="iscsi.go:774"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> GetIfaces" file="iscsi.go:511"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> GetIscsiIfacesPath" file="iscsi.go:492"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> FileExists for path /var/lib/iscsi/ifaces" file="file.go:62"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< FileExists" file="file.go:68"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> FileExists for path /etc/iscsi/ifaces" file="file.go:62"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< FileExists" file="file.go:68"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< GetIscsiIfacesPath" file="iscsi.go:499"
time="2019-12-20T00:44:18Z" level=trace msg="no iscsi ifaces found" file="iscsi.go:524"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< GetIfaces" file="iscsi.go:525"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> loginToTarget with targetIqn: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2" file="iscsi.go:292"
time="2019-12-20T00:44:18Z" level=trace msg="Checking with TargetName: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2" file="iscsi.go:295"
time="2019-12-20T00:44:18Z" level=debug msg="Found target :Target:iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2" file="iscsi.go:297"
time="2019-12-20T00:44:18Z" level=trace msg=">>>>> addTarget called with target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2 address: 172.16.226.158 port: 3260" file="iscsi.go:446"
time="2019-12-20T00:44:18Z" level=trace msg="ExecCommandOutputWithTimeout called with iscsiadm[--mode node --targetname iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2 --login] 60" file="cmd.go:35"
time="2019-12-20T00:44:18Z" level=trace msg="process with pid: 29 finished successfully" file="cmd.go:62"
time="2019-12-20T00:44:18Z" level=trace msg="Logging in to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.34.158,3260] (multiple)\nLogging in to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.226.158,3260] (multiple)\nLogin to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.34.158,3260] successful.\nLogin to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.226.158,3260] successful.\n" file="cmd.go:66"
time="2019-12-20T00:44:18Z" level=trace msg="addTarget Response :Logging in to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.34.158,3260] (multiple)\nLogging in to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.226.158,3260] (multiple)\nLogin to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.34.158,3260] successful.\nLogin to [iface: default, target: iqn.2007-11.com.nimblestorage:group-skrishna-int-va-g66c9b39533bd05d2, portal: 172.16.226.158,3260] successful.\n" file="iscsi.go:482"
time="2019-12-20T00:44:18Z" level=trace msg="<<<<< addTarget" file="iscsi.go:487"
```

**Pull request merged**